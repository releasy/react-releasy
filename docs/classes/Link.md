# Link

A class responsible to fetch data from server.

## Usage

```javascript
import { Link } from 'react-releasy';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
  fetchTimeout: 30000,
  retryDelays: [1000, 3000, 5000, 10000],
  headers: {},
});
```

Also, we can use a function to generate the `headers` object:

```javascript
import { Link } from 'react-releasy';

// this function would get the token inside localStorage or AsyncStorage
import { getToken } from './auth';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
  fetchTimeout: 30000,
  retryDelays: [1000, 3000, 5000, 10000],
  headers: () => ({ token: getToken() }),
});
```

## Props

- **url**: the server address
- **fetchTimeout**: how long to wait(in ms) for a success response *(optional)*
- **retryDelays**: how long to wait(in ms) before make a new request after a failure *(optional)*
- **headers**: additional headers that will be send to the server *(optional)*

----

[Go back](../)
