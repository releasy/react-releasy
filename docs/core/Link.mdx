---
name: Link
menu: Core
route: /core/link
---

# Link

A class responsible to fetch data from server.

## Usage

```javascript
import { Link } from 'react-releasy';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
  fetchTimeout: 30000,
  retryDelays: [1000, 3000, 5000, 10000],
  headers: {},
});
```

Also, we can use a function to generate the `headers` object:

```javascript
import { Link } from 'react-releasy';

// this function would get the token inside localStorage or AsyncStorage
import { getToken } from './auth';

const link = new Link({
  url: 'https://yourserveraddress.com/graphql',
  fetchTimeout: 30000,
  retryDelays: [1000, 3000, 5000, 10000],
  headers: () => ({ token: getToken() }),
});
```