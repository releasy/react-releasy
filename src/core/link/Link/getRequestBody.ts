import { RequestNode, UploadableMap, Variables } from 'relay-runtime';

const getRequestBodyWithUploadables = (
  request: RequestNode,
  variables: Variables,
  uploadables: UploadableMap,
) => {
  const formData = new FormData();
  formData.append('query', request.text);
  formData.append('variables', JSON.stringify(variables));

  Object.keys(uploadables).forEach((key) => {
    formData.append(key, uploadables[key]);
  });

  return formData;
};

const getRequestBodyWithoutUplodables = (request: RequestNode, variables: Variables): string => {
  return JSON.stringify({
    variables,
    query: request.text,
  });
};

const getRequestBody = (
  request: RequestNode,
  variables: Variables,
  uploadables?: UploadableMap,
): FormData | string => {
  if (uploadables) {
    return getRequestBodyWithUploadables(request, variables, uploadables);
  }

  return getRequestBodyWithoutUplodables(request, variables);
};

export default getRequestBody;
