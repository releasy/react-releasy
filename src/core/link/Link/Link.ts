import { CacheConfig, RequestNode, UploadableMap, Variables, GraphQLResponse } from 'relay-runtime';
import invariant from 'invariant';
import fetchWithRetries from 'fbjs/lib/fetchWithRetries';

import { LinkInterface } from '../../Config';
import getHeaders from './getHeaders';
import getRequestBody from './getRequestBody';

export type LinkPropsType = {
  url: string,
  fetchTimeout?: number,
  retryDelays?: number[],
  headers?: Object | Function,
};

class Link implements LinkInterface {
  url: string;
  fetchTimeout: number;
  retryDelays: number[];
  headers: Object | Function;

  constructor(props: LinkPropsType) {
    invariant(props, 'Link: missing props');
    invariant(props.url, 'Link: missing "url"');

    this.url = props.url;
    this.headers = props.headers || {};
    this.fetchTimeout = props.fetchTimeout || 30000;
    this.retryDelays = props.retryDelays || [1000, 3000, 5000, 10000];
  }

  fetch = (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<GraphQLResponse> => {
    return new Promise(async (resolve) => {
      const customHeaders = typeof this.headers === 'function' ? this.headers() : this.headers;

      const headers = {
        ...getHeaders(uploadables),
        ...customHeaders,
      };

      const body = getRequestBody(request, variables, uploadables);

      const response = await fetchWithRetries(this.url, {
        headers,
        body,
        method: 'POST',
        fetchTimeout: this.fetchTimeout,
        retryDelays: this.retryDelays,
      });

      const contentType = response.headers.get('content-type');
      if (contentType && contentType.indexOf('application/json') !== -1) {
        resolve(response.json());
        return;
      }

      resolve(response.text());
    });
  }
}

export default Link;
