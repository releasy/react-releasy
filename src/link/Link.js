// @flow
import invariant from 'invariant';
import fetchWithRetries from 'fbjs/lib/fetchWithRetries';

import type { CacheConfig, RequestNode, UploadableMap, Variables } from 'relay-runtime';

import getHeaders from './getHeaders';
import getRequestBody from './getRequestBody';

export type Props = {
  url: string,
  fetchTimeout?: number,
  retryDelays?: Array<number>,
  headers?: Object | () => Object,
};

class Link {
  url: string;
  fetchTimeout: number;
  retryDelays: Array<number>;
  headers: Object | () => Object;

  constructor(props: Props) {
    invariant(props, 'Link: missing props');
    invariant(props.url, 'Link: missing "url"');

    this.url = props.url;
    this.headers = props.headers || {};
    this.fetchTimeout = props.fetchTimeout || 30000;
    this.retryDelays = props.retryDelays || [1000, 3000, 5000, 10000];
  }

  fetch = async (
    request: RequestNode,
    variables: Variables,
    cacheConfig: CacheConfig,
    uploadables?: UploadableMap,
  ): Promise<any> => {
    const customHeaders = typeof this.headers === 'function' ? this.headers() : this.headers;

    const headers = {
      ...getHeaders(uploadables),
      ...customHeaders,
    };

    const body = getRequestBody(request, variables, uploadables);

    const response = await fetchWithRetries(this.url, {
      method: 'POST',
      fetchTimeout: this.fetchTimeout,
      retryDelays: this.retryDelays,
      headers,
      body,
    });

    const contentType = response.headers.get('content-type');
    if (contentType && contentType.indexOf('application/json') !== -1) {
      return response.json();
    }

    return response.text();
  }
}

export default Link;
