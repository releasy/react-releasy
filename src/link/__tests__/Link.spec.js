import mockFetch from '../../../test/fixtures/fetch/mockFetch';
import createCacheConfig from '../../../test/fixtures/relay/createCacheConfig';
import createRequest from '../../../test/fixtures/relay/createRequest';
import createUploadables from '../../../test/fixtures/relay/createUploadables';
import createVariables from '../../../test/fixtures/relay/createVariables';
import { CONTENT_TYPE, MOCK_TYPE } from '../../../test/consts';

import Link from '../Link';

it('should create a valid Link', () => {
  const link = new Link({
    url: '/graphql',
  });

  expect(link).toMatchSnapshot();
});

it('should thrown an error without props', () => {
  const createLink = () => {
    new Link();
  };

  expect(createLink).toThrowErrorMatchingSnapshot();
});

it('should thrown an error without "url"', () => {
  const createLink = () => {
    new Link({});
  };

  expect(createLink).toThrowErrorMatchingSnapshot();
});

it('should fetch default data', async () => {
  mockFetch();

  const link = new Link({
    url: '/graphql',
  });

  const request = createRequest();
  const variables = createVariables();
  const cacheConfig = createCacheConfig();

  const data = await link.fetch(request, variables, cacheConfig);

  expect(data).toMatchSnapshot();
});

it('should fetch data with uploadables', async () => {
  mockFetch({ contentType: CONTENT_TYPE.ALL, mock: MOCK_TYPE.UPLOADABLE_MUTATION });

  const link = new Link({
    url: '/graphql',
  });

  const request = createRequest(MOCK_TYPE.UPLOADABLE_MUTATION);
  const variables = createVariables();
  const cacheConfig = createCacheConfig();
  const uploadables = createUploadables();

  const data = await link.fetch(request, variables, cacheConfig, uploadables);

  expect(data).toMatchSnapshot();
});
