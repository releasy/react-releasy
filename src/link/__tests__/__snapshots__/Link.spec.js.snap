// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create a valid Link 1`] = `
Link {
  "fetch": [Function],
  "fetchTimeout": 30000,
  "headers": Object {},
  "retryDelays": Array [
    1000,
    3000,
    5000,
    10000,
  ],
  "url": "/graphql",
}
`;

exports[`should fetch data with uploadables 1`] = `""`;

exports[`should fetch default data 1`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "fakeViewerId",
      "name": "John Doe",
    },
  },
}
`;

exports[`should thrown an error without "url" 1`] = `"Fetch: missing \\"url\\""`;

exports[`should thrown an error without props 1`] = `"Fetch: missing props"`;

exports[`should use custom headers 1`] = `
Object {
  "data": Object {
    "me": Object {
      "id": "fakeViewerId",
      "name": "John Doe",
    },
  },
}
`;
